### ISSUE TITLE
Expose public API for manipulating the virtual filesystem

### ISSUE THREAD

**Issue by rfk** [rfk/pypyjs#132]

**Comment by jedie** [comment]
Is there no deep hock for file access?!?

So it will be possible to so this
* **__import__("foo")**
* catch the access
* pause VM
* get the file content via AJAX
* insert it to virtual filesystem
* resume VM

So the import errors like the followed, will fly away:
* #137 
* #109 

Because we didn't need the current "importStatementRE"-work-a-round

**Comment by rfk** [comment]
Indeed not, because currently there's no support for this step:

> * pause VM

If we manage the figure that out, by e.g. using emscripten's new "emterpreter" feature, then a lot of nice things become possible.

**Comment by traverseda** [comment]
Perhaps

```
import pathedImportLib as import
import("foo")
patched importlib uses jsbridge to get content via AJAX 
```

or

```
import virtualFile
fooFile = virtualFile('foo.py' ,'w+')

import urllib2 #I wish. Maybe something can be done with websockets, but I'm not holding my breath.
fooFile.write(urllib2.urlopen("http://localhost/py/foo.py"))
fooFile.save()

import foo
```

Or something like that, but that would actaully work. The point is to try and keep things nice and pythonic. We shouldn't need to pause the VM, bacause we should be able to make that request in the VM.

**Comment by jedie** [comment]
@rfk :
> Indeed not, because currently there's no support for this step:
>>  pause VM

Hm! If we patch the general file access routines deep inside pypy, then we didn't really need a "pause VM"... Accessing a file will also block and didn't has a timeout, isn't it?
So replace **open()** with a AJAX routine will work?

@traverseda :
> Or something like that, but that would actaully work. 

But this is more a temporary work-a-round... Maybe this is easier to implement.

**Comment by rfk** [comment]
> So replace open() with a AJAX routine will work?

Not really.  The problem is that in general AJAX requests from JS are asynchronous [1], and you can only really get the results via a callback.  If we tried to implement an ajax-file-fetching routine from python, it'd have to be done using callbacks like this:

```
import js
jquery = js.globals['$']

def fetch_file_via_ajax(url, outpath, callback):
    def on_success(data):
        with open(outpath, "w") as f:
            f.write(data)
        callback()

    jquery.ajax({
        "url": url,
        "method": "GET",
        "success": on_success
    })
```

You can't transparently replace a synchronous call like `import foo` with an callback-driven approach like the above.  Unless you can pause and resume the VM.

You *can* do synchronous AJAX requests if you're in a WebWorker, and the approach described above would work fine there.  In fact, I believe emscripten has support for a transparently-fetch-files-via-ajax-on-demand filesystem that can be used inside a worker.  But this approach won't work when running in the main page.

[1] https://developer.mozilla.org/en-US/docs/Web/API/XMLHttpRequest/Synchronous_and_Asynchronous_Requests#Synchronous_request

**Comment by jedie** [comment]
HM. Boring ;(
But this will be the best solution, isn't it?

**Comment by jedie** [comment]
from https://github.com/pypyjs/pypyjs/issues/122#issuecomment-115648709 :

```
with open("/lib/pypyjs/lib_pypy/my_module.py", "w") as fl:
     fl.write("x='hi'")

import my_module

print my_module.x # hi
print "my_module.__file__:", my_module.__file__

import os
for filename in os.listdir("/lib/pypyjs/lib_pypy/"):
    if filename.startswith("my"):
        print "*", filename
```
output:
```
hi
my_module.__file__: /lib/pypyjs/lib_pypy/my_module.py
* my_module.py
* my_module.pyc
```

This is IMHO a interesting feature! Manipulating the virtual filesystem is easy from python...


So, somthing like the following pseudo code, can be possible?
```
import js

js.request("my_app_package.tar.gz") # maybe: .egg or .wheel?

import my_app

print my_app.__file__ # -> /lib/pypyjs/lib_pypy/my_app/__init__.py
```

**Comment by ShaneFarrell** [comment]
How are you polling stdin? Because whatever method you're using for that doesn't seem to be destroying performance. If it's some kind of busy loop and that's not crashing the page then I don't see why a busy loop can't be used to suspend the vm.

**Comment by rfk** [comment]
There's no polling of stdin, the terminal widget does an async prompt to gather input from the user and push it into the vm.

**Comment by ShaneFarrell** [comment]
Since the state of the vm isn't being reset between each call I assume that means that when the vm runs out of code to run it becomes inactive until it is given more code. Therefore maybe you could run the entire program inside a python generator function (i.e. a green thread). When the program wants to block you can yield from this generator and let the vm fall off the end of the code to run. Then whenever you want to restart the vm you could pass in code that called next on the iterator in order to restart it.

```python

def main():
    print "before"
    yield
    print "after"

runner = main()

runner.next()
runner.next() # Set up a JS timeout to send you this after a delay
```

**Comment by ShaneFarrell** [comment]
Actually this might be worth looking at given my previous comment: http://pypy.readthedocs.org/en/latest/stackless.html

I don't have time to look at it right now, but it sounds like pypy has green threading built in. In which case you might be able to do the trick above.

**Comment by ShaneFarrell** [comment]
Having a JS way to access the file system would be really useful to me now.

**Comment by Darker** [comment]
Well is it at least possible to put the files in pypy emulator through stdin?
